xls/examples/hack.x

```diff
@@ -0,0 +1,258 @@
+    (value,)
+}
+
+#![test]
+fn decode_a_test() {
+    // @21
+    let (value) = decode_a_instruction(u16:0b0_000000000010101);
+    let _ = assert_eq(value, u15:21);
+    _
+}
+
+const COMP_0 = u7:0b0101010;
+const COMP_1 = u7:0b0111111;
+const COMP_MINUS_1 = u7:0b0111010;
+const COMP_D = u7:0b0001100;
+const COMP_A = u7:0b0110000;
+const COMP_M = u7:0b1110000;
+const COMP_NOT_D = u7:0b0001101;
+const COMP_NOT_A = u7:0b0110001;
+const COMP_NOT_M = u7:0b1110001;
+const COMP_MINUS_D = u7:0b0001111;
+const COMP_MINUS_A = u7:0b0110011;
+const COMP_MINUS_M = u7:0b1110011;
+const COMP_D_PLUS_1 = u7:0b0011111;
+const COMP_A_PLUS_1 = u7:0b0110111;
+const COMP_M_PLUS_1 = u7:0b1110111;
+const COMP_D_MINUS_1 = u7:0b0001110;
+const COMP_A_MINUS_1 = u7:0b0110010;
+const COMP_M_MINUS_1 = u7:0b1110010;
+const COMP_D_PLUS_A = u7:0b0000010;
+const COMP_D_PLUS_M = u7:0b1000010;
+const COMP_D_MINUS_A = u7:0b0010011;
+const COMP_D_MINUS_M = u7:0b1010011;
+const COMP_A_MINUS_D = u7:0b0000111;
+const COMP_M_MINUS_D = u7:0b1000111;
+const COMP_D_AND_A = u7:0b0000000;
+const COMP_D_AND_M = u7:0b1000000;
+const COMP_D_OR_A = u7:0b0010101;
+const COMP_D_OR_M = u7:0b1010101;
+
+const DEST_NULL = u3:0b000;
+const DEST_M = u3:0b001;
+const DEST_D = u3:0b010;
+const DEST_A = u3:0b100;
+
+const JUMP_NULL = u3:0b000;
+const JUMP_JGT = u3:0b001;
+const JUMP_JEQ = u3:0b010;
+const JUMP_JGE = u3:0b011;
+const JUMP_JLT = u3:0b100;
+const JUMP_JNE = u3:0b101;
+const JUMP_JLE = u3:0b110;
+const JUMP_JMP = u3:0b111;
+
+fn decode_c_instruction(ins: u16) -> (u7, u3, u3) {
+    let comp: u7 = ins[6+:u7];
+    let dest: u3 = ins[3+:u3];
+    let jump: u3 = ins[0+:u3];
+    (comp, dest, jump)
+}
+
+#![test]
+fn decode_c_test() {
+    // MD=D+1
+    let (comp, dest, jump) = decode_c_instruction(u16:0b1110011111011000);
+    let _ = assert_eq(comp, COMP_D_PLUS_1);
+    let _ = assert_eq(dest, DEST_M|DEST_D);
+    let _ = assert_eq(jump, JUMP_NULL);
+    // M=1
+    let (comp, dest, jump) = decode_c_instruction(u16:0b111_0111111_001_000);
+    let _ = assert_eq(comp, COMP_1);
+    let _ = assert_eq(dest, DEST_M);
+    let _ = assert_eq(jump, JUMP_NULL);
+    // D+1;JLE
+    let (comp, dest, jump) = decode_c_instruction(u16:0b111_0011111_000_110);
+    let _ = assert_eq(comp, COMP_D_PLUS_1);
+    let _ = assert_eq(dest, DEST_NULL);
+    let _ = assert_eq(jump, JUMP_JLE);
+    _
+}
+
+fn run_a_instruction(pc: u16, ins: u16, rd: u16, ra: u16, rm: u16) -> (u16, u16, u16, u16, u1) {
+    let (value) = decode_a_instruction(ins);
+    let ra' = value as u16;
+    let wm = u1:0;
+    (pc + u16:1, rd, ra', rm, wm)
+}
+
+#![test]
+fn run_a_test() {
+    // @21
+    let (pc, rd, ra, rm, wm) = run_a_instruction(u16:0, u16:0b0_000000000010101, u16:1, u16:2, u16:3);
+    let _ = assert_eq(pc, u16:1);
+    let _ = assert_eq(rd, u16:1);
+    let _ = assert_eq(ra, u16:21);
+    let _ = assert_eq(rm, u16:3);
+    let _ = assert_eq(wm, u1:0);    
+    _
+}
+
+fn z_bit(z: u1, v: u16) -> u16 {
```

Comment:
To be consistent, can these functions can be tested?

Reply:
Done